python/pyarrow/includes/libarrow_dataset.pxd

```diff
@@ -0,0 +1,365 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# distutils: language = c++
+
+from libcpp.unordered_map cimport unordered_map
+
+from pyarrow.includes.common cimport *
+from pyarrow.includes.libarrow cimport *
+
+
+cdef extern from "arrow/api.h" namespace "arrow" nogil:
+
+    cdef cppclass CRecordBatchIterator "arrow::RecordBatchIterator"(
+            CIterator[shared_ptr[CRecordBatch]]):
+        pass
+
+
+cdef extern from "arrow/dataset/api.h" namespace "arrow::fs" nogil:
+
+    ctypedef shared_ptr[CFileSystem] CFileSystemPtr "arrow::fs::FileSystemPtr"
```

Comment:
What's the current plan with the `Ptr` aliases?

Reply:
I'm using the forward declared aliases (*Ptr, *Vector, *Iterator) mainly in the libarros_*.pxd header definitions to keep it a bot closer to the C++ header files. In the *.pyx files I tend spell out the types, being more explicit.

Reply:
The `Ptr` aliases are slated for removal soonish; in a follow up it'd be good to remove usages of them in cython to alleviate breakage when we do the same in C++